Lab1_Operationalizing_Pytorch_with_Mlflow/LAB.md

```diff
@@ -29,5 +24,101 @@ itself is organized into subfolders according to which components it belongs to;
+| 8 | Scheduler | Some activities are triggered by clock | All Nodes and Machines | CRON | We are using Unix like machines |
+| 9 | Project Metadata | We need to describe the ML Component, its structure and how it should run | Source Code Repository<br>Training and Serving Nodes | MLFlow Project Definition | Non-functional Requirement we must adhere |
+| 10 | ML Lifecycle Management | We must track the ML Process | All Nodes and Machines | MLFlow |  Non-functional Requirement we must adhere |
+| 11 | ML Framework | We are building ML Components. Therefore, we must use a ML framework to train our models | Training and Serving Nodes | Pytorch | Non-functional Requirement we must adhere |
+
+
+### Workflow
+
+[DevOps](https://en.wikipedia.org/wiki/DevOps) tools heavily relies on actions triggered by changes
+and other events regarding application's source code. MLOps goes beyond this assumption since any reasonable
+solution must deal with assets like ML Models, datasets, features, etc. Therefore, a simples ML 
+application may turn into a complex workflow in order to ensure the best model runs in Production. The
+following workflow depicts a common development scenario where a production web application uses a ML
+Model to provide business value. 
+
+![Solution Workflow](imgs/workflow.svg)
+
+The first stage is Development Workflow. It concerns the continuous integration of new ML code and Dataset 
+changes into a Development Server. Here follows the steps:
+
+1. Commited code or changes in the Dataset fires a training execution. These events trigger Github actions;
+ 
+ 1. ML pipeline starts. we are considering that previous data activities (Extraction, Validation and
+ Preparation) were already executed. In _Training Node_, a Github runner checks out the ML Component 
+ and \ or updates the training dataset. The ML Component itself describes its structure (_MLFlow project file_) 
+ and how it should be run (_Dockerfile_). The runner starts the training process using MLFlow CLI 
+ features: ```mlflow run```. The training happens in a GPU-powered Docker container configured with the 
+ _Tracking Server_ (```MLFLOW_TRACKING_URI```) . At the end of this step, there's a new trained model 
+ with its respective metadata;
+ 
+ 1. The Training Node sends the trained model, its extra files, and metadata to the Model Registry
+ and Tracking Server respectively. In fact, these assets are physically stored in an external Cloud
+ Storage Service acessible from Internet;
+ 
+ 1. The development workflow ends by deploying the trained model in a _Serving Node_. The runner creates
+ and registers a new Docker image capable running the model. The runner starts a container  
+ (```docker run```) which downloads and packs the trained model through command 
+ ```mlflow create deployment```.
+
+In our scenario, the decision to go to Production must be triggered through MLFlow Tracking Server. Here
```

Comment:
This could potentially be done through a "release" commit in GitHub, precipitating an Action that performs this workflow? 
I think you can change the stage of a model through the CLI, so a release commit could indicate the model number to move to production, then runs an ```mlflow create deployment```
OR
we ignore the mlflow tagging of models and ship the docker image we're happy with to production

Reply:
First of all, Sorry for the long text. Got excited with the topic. :-P

I was in a hurry to finish this yesterday and just wrote it down as it came out of my mind. However, it's is really tricky (and funny). We should discuss about it. IMHO, There are some points we gotta decide:

**1. Which tool should we use to trigger the deployment process (to Production)?**
When I proposed the MLFlow should be our guy for that, I was thinking about putting it in a "nice spot" and bring relevance to the tool. However, it seems it isn't ready for that: https://github.com/mlflow/mlflow/issues/3015 . At the same time, (Probably) GitHub Action can do that "hands down". Going for MLFLow would demand us to incorporate the concept of events and plug another structure in our architecture (Kafka, ZeroMQ or antoher message bus). It would be a huge contribution... and risky as well. It's a long shot.  Being pragmatic, we should use GItHub Actions because we don't have time anymore. Nevertheless, I can't help thinking about how nice it would be if somehow we integrate MLFlow and an event bus. :-P

**2. How strict about rules like "Promote Packages" and "Build Once" should we be?**
I guess it's all about what we are considering as deliverables. If the model is our deliverable, we go for the model number. This implies that moving to production also means creating a new Docker Image. I like the idea that the whole image is our deliverable. This strengthens so many DevOps concepts. Can you think in a simple way to link the model and its serving docker image (besides naming the image with the model id)?

Reply:
There's a github action that can query MLFlow and compare a dev model against prod one: https://github.com/marketplace/actions/mlflow-tracking

We might use this to promote best model to prod, but this requires some extra work for defining queries and the workflow itself. This could be an area for future development on our demo, for instance.

Reply:
> There's a github action that can query MLFlow and compare a dev model against prod one: https://github.com/marketplace/actions/mlflow-tracking
> 

Cool. Poll for updates and go for releasing. We still have to decide which one will be our deliverable to Production: the model itself or the Serving image. The first implies we gotta build the image again (maybe breaking the rules of "build only once" and  "you deploy what you test"). The second implies we must associate model and image (also manage it in a Image Registry) so when one promotes a model, its serving image is deployed.

Reply:
That's a good question. I was planning to embed model within the serving image, but I don't mind another approach as well.

Reply:
> That's a good question. I was planning to embed model within the serving image, but I don't mind another approach as well.

Yes. I like this way. But how do we associate the promoted model with that image? I mean the very image we have built and registered during Dev workflow. I was thinking about a naming pattern: `MLCOMPONENT_MODELID` or something like that. How does it sound to you guys?

Reply:
I'm not sure I follow. In my understanding every serving image contains associated model — so we have a bunch of docker images with torchserve + model, each of them ready to be deployed to prod. We do need to specify which of the images is "current" (i.e. should be deployed as production) and I don't have a ready answer for that, but other than that, what do we need the naming pattern for?

I certainly might be missing something in the workflow, so I feel free to correct me :)

Reply:
Actually, It seems you are right. :smile:  There is no need for a complex naming pattern at all. It's much more simple than that. Realized after reading again MLFLow docs. I would just change one premise: We may deploy to Prod whatever model version we want. Following John's ideas and yours:

1. GitHub Actions (GA) controls the flow and MLFlow holds proper information regarding models
1. We're in Dev Workflow. Model has just been trained and registered. We embed it in a image tagged with `-t model:version`
1. The model continues to evolve and many executions later we decide to release version _X_
1. We issue a release commit (or something like that) informing the version. GA triggers the jobs and send commands to Production Servers
1. GA (using CMD or REST API) updates model information in MLFlow finally promoting version _X_ to Production Stage.

What do you think? Am I overcomplicating things? :eyes:

Reply:
I've tried to put your suggestions in a new version. Check it out.